### PR TITLE
Complete schema for Antsibull changelogs

### PR DESCRIPTION
**Pull request by shatakshiiii**
Related: #2814

### CODE REVIEW

**Comment by ssbarnea** [comment]
@felixfontein Any chance you can check this one? Is the new scheme working as expected?

**Review by ssbarnea — CHANGES_REQUESTED** [review]
Few observations remarked by @felixfontein 👍 
- missing known_issues
- missing trivial category
- I see various markdownDescription fields containing a leading space
- "Value of `short_description from plugin `DOCUMENTATION`." - is missing one backquote
- a lot of duplication, fields that are the same should never be duplicated inside the file and references should be used instead.

File: src/ansiblelint/schemas/changelog.json [hunk 1]
```diff
@@ -18,31 +46,69 @@
...
+              "type": "array"
+            },
             "minor_changes": {
               "items": {
                 "type": "string"
               },
               "type": "array"
             },
             "release_summary": {
+              "markdownDescription": "This must be valid [reStructuredText](https://en.wikipedia.org/wiki/ReStructuredText).",
```
Comment (felixfontein) [inline src/ansiblelint/schemas/changelog.json:68]: This also applies to every list entry in the other sections.

File: src/ansiblelint/schemas/changelog.json [hunk 2]
```diff
@@ -57,6 +123,100 @@
...
           "type": "array"
         },
+        "objects": {
+          "additionalProperties": false,
+          "properties": {
+            "playbook": {
+              "items": {
+                "properties": {
+                  "description": {
+                    "markdownDescription": "Value of `short_description` from plugin `DOCUMENTATION`.",
```
Comment (felixfontein) [inline src/ansiblelint/schemas/changelog.json:133]: Playbooks are not documentable yet.
```suggestion
                    "markdownDescription": "A short description of what the playbook does.",
```

File: src/ansiblelint/schemas/changelog.json [hunk 3]
```diff
@@ -57,6 +123,100 @@
...
+                },
+                "type": "object"
+              },
+              "type": "array"
+            },
+            "role": {
+              "items": {
+                "properties": {
+                  "description": {
+                    "markdownDescription": "Value of `short_description` from plugin `DOCUMENTATION`.",
```
Comment (felixfontein) [inline src/ansiblelint/schemas/changelog.json:155]:
```suggestion
                    "markdownDescription": "Value of `short_description` from role's argument spec.",
```

File: src/ansiblelint/schemas/changelog.json [hunk 4]
```diff
@@ -57,6 +123,100 @@
...
+            "callback": {
+              "$ref": "#/$defs/plugin-descriptions"
+            },
+            "cliconf": {
+              "$ref": "#/$defs/plugin-descriptions"
+            },
+            "connections": {
+              "$ref": "#/$defs/plugin-descriptions"
+            },
+            "filter": {},
```
Comment (felixfontein) [inline src/ansiblelint/schemas/changelog.json:194]: Why is this empty?

File: src/ansiblelint/schemas/changelog.json [hunk 5]
```diff
@@ -57,6 +123,100 @@
...
+            "netconf": {
+              "$ref": "#/$defs/plugin-descriptions"
+            },
+            "shell": {
+              "$ref": "#/$defs/plugin-descriptions"
+            },
+            "strategy": {
+              "$ref": "#/$defs/plugin-descriptions"
+            },
+            "test": {},
```
Comment (felixfontein) [inline src/ansiblelint/schemas/changelog.json:213]: Same question here.

File: src/ansiblelint/schemas/changelog.json [hunk 6]
```diff
@@ -70,15 +230,14 @@
     "semver": {
       "pattern": "\\d+.\\d+.\\d+.*",
```
Comment (felixfontein) [inline src/ansiblelint/schemas/changelog.json:231]: See "Is there a suggested regular expression (RegEx) to check a SemVer string?" on https://semver.org/

**Review by felixfontein — COMMENTED** [review]
Line 119 should be
```
                "type": "string"
```
`null` isn't allowed for modules. (I cannot add a comment to that line, so writing it here.)

File: src/ansiblelint/schemas/changelog.json (hunk 4 above)
Comment (ssbarnea) [reply src/ansiblelint/schemas/changelog.json:194]: The doc did not give any details regarding what filter allows.

File: src/ansiblelint/schemas/changelog.json (hunk 5 above)
Comment (ssbarnea) [reply src/ansiblelint/schemas/changelog.json:213]: Same here.

File: src/ansiblelint/schemas/changelog.json (hunk 6 above)
Comment (ssbarnea) [reply src/ansiblelint/schemas/changelog.json:231]: JSON specification does only allow very limited set of regex. https://json-schema.org/understanding-json-schema/reference/regular_expressions.html -- very limited!

File: src/ansiblelint/schemas/changelog.json (hunk 1 above)
Comment (ssbarnea) [reply src/ansiblelint/schemas/changelog.json:68]: I am not going to copy the text in each section. There is no way to inherit it.

File: src/ansiblelint/schemas/changelog.json (hunk 4 above)
Comment (felixfontein) [reply src/ansiblelint/schemas/changelog.json:194]: Exactly the same as every other plugin type.

File: src/ansiblelint/schemas/changelog.json (hunk 1 above)
Comment (felixfontein) [reply src/ansiblelint/schemas/changelog.json:68]: I would have assumed you could do something like `"$ref": "#/$defs/changelog-section"` (for a generic 'list' section).

File: src/ansiblelint/schemas/changelog.json [hunk 7]
```diff
@@ -18,45 +46,182 @@
...
           "type": "object"
         },
         "codename": {
           "type": "string"
         },
+        "fragments": {
+          "items": {
+            "type": "string"
+          },
+          "markdownDescription": "List of strings representing filenames with framents.",
```
Comment (felixfontein) [inline src/ansiblelint/schemas/changelog.json:100]:
```suggestion
          "markdownDescription": "List of strings representing filenames of changelog fragments.",
```

File: src/ansiblelint/schemas/changelog.json [hunk 8]
```diff
@@ -18,45 +46,182 @@
...
               "name": {
-                "markdownDescription": " It must not be the FQCN, but the name inside the collection.",
+                "markdownDescription": "It must not be the FQCN, but the name inside the collection.",
                 "pattern": "[a-zA-Z0-9_]+",
                 "title": "Short module name",
                 "type": "string"
               },
               "namespace": {
-                "markdownDescription": "Must be `null` for plugin and objects. For modules it must be either empty string. The namespace is used to group new modules by their namespace inside the collection.",
+                "markdownDescription": "Must be `''` for modules directly in `plugins/modules/`, or the dot-separated list of directories the module is in inside the `plugins/modules/` directory. This is mostly relevant for large collections such as `community.general` and `community.network`. For example, the `community.general.docker_container` module is in the directory `plugins/modules/cloud/docker/`, hence its namespace must be `cloud.docker`. The namespace is used to group new modules by their namespace inside the collection.",
```
Comment (felixfontein) [inline src/ansiblelint/schemas/changelog.json:118]: docker_container hasn't been in community.general for years, and community.general no longer has hat mentioned directory structure for two months. (I see where you copied this from, I guess I have to adjust that description :) )

File: src/ansiblelint/schemas/changelog.json [hunk 9]
```diff
@@ -18,45 +46,182 @@
...
                 "title": "Namespace",
-                "type": ["string", "null"]
+                "type": ["string"]
```
Comment (felixfontein) [inline src/ansiblelint/schemas/changelog.json:120]: Shouldn't this be
```suggestion
                "type": "string"
```
? (I guess both work, but this is the syntax you use in other cases.)

File: src/ansiblelint/schemas/changelog.json (hunk 8 above)
Comment (felixfontein) [reply src/ansiblelint/schemas/changelog.json:118]: Ref: https://github.com/ansible-community/antsibull-changelog/pull/95

**Review by felixfontein — APPROVED** [review]

**Comment by felixfontein** [comment]
@shatakshiiii @ssbarnea thanks for improving the schema!

**Pull request merged**